Batched

Tom-Willemsen opened the pull request:
~5x perf

rerpha commented on tsconfig.json:
```diff
@@ -13,6 +13,7 @@
     "jsx": "preserve",
     "incremental": true,
     "target": "ES2023",
+    "downlevelIteration": true,
```
I don't know if you need this if you're using a new enough target - i think webstorm/pycharm is being silly

rerpha commented on next.config.ts:
```diff
@@ -6,6 +6,15 @@ const nextConfig: NextConfig = {
   images: {
     unoptimized: true,
   },
+  // compress: false,
+  // webpack(webpackConfig) {
+  //   return {
+  //     ...webpackConfig,
+  //     optimization: {
+  //       minimize: false,
+  //     },
+  //   };
+  // },
```
does the prod build actually use webpack now rather than turbopack?

Tom-Willemsen commented:
Here’s a detailed code review of PR #156 for ISISComputingGroup/WebDashboard:

---

## Overview

This PR refactors the internal representation of blocks, groups, and dashboard data from Array-based to Map-based data structures. It touches a wide range of files, including major React components, utility functions, and test files. The stated goal appears to be improving lookup efficiency and simplifying certain operations.

## Key Changes
1. Switch from Arrays to Maps
  - Types: New types (tBlockMapping, tGroups) are introduced in types.ts to define Maps for block and group data.
  - Removal of IfcGroup and DashboardArr: The old group and dashboard array types are removed, and related code is updated to use Maps.
  - Component Props/State: All relevant components (e.g., Group, Groups, Instrument, TopBar) now accept and operate on Maps rather than arrays.
  - Instrument Class: Refactored to initialize and use Maps for groups, dashboard, and runInfoPVs.
2. Component and Utility Updates
  - Group & Groups Components: Updated to iterate over Maps using Array.from(map.entries()) or Array.from(map.values()).
  - Block Component: No direct change to Map/Array logic, but minor optimizations to the UI update logic for value changes.
  - InstrumentData: Major refactor to batch process websocket messages, avoid direct state mutation, and handle updates via Map-based data structures.
  - TopBar Component: Updated to use Maps for the dashboard and run info PVs, and revised the dashboard rendering accordingly.
  - GroupUtils and PVutils: Refactored helper/utilities to use Maps and removed now-unnecessary findPVByAddress and findPVByHumanReadableName functions.
3. Test Suite Updates
  - All affected test files are updated to work with Map-based test data and to remove references to the old types and helper functions.
  - Snapshots: Updated to reflect small but relevant UI changes.

## Strengths
  - Performance: Using Maps for lookup-heavy structures (blocks, groups, PVs) is a sound optimization, especially for larger datasets.
  - Consistency: The refactor is applied consistently throughout the codebase, including tests and snapshots.
  - Batching/Cloning: The new approach in InstrumentData for batching websocket updates and cloning state before mutation is more React-friendly.
  - Type Safety: Introducing new types (e.g., tBlockMapping, tGroups) improves clarity and type safety.
  - Component Memoization: Use of React.memo for pure functional components is a good performance optimization.
  - Code Cleanup: Removal of obsolete utilities (e.g., various findPV... helpers) reduces maintenance overhead.

## Weaknesses / Suggestions

  - Map as State and Props
    * React state and props containing Maps can cause subtle bugs because Maps are reference types and React’s shallow comparison may not detect changes. The code often uses structuredClone() when cloning state, but this isn’t always cheap, and may not be supported in all environments. Consider using plain objects if possible, or ensure Map usage is always safe and compatible with your state management approach.

  - UI and Accessibility
    * Some component UI code (e.g., in TopBar) has become more complex due to manual table construction. Ensure that accessibility (ARIA roles, keyboard navigation) is not negatively affected by the refactor.
        Consider extracting dashboard row/column rendering into small subcomponents for readability.
  - Error Handling
        The websocket message queue and interval batching is robust, but failure modes (e.g., missed updates, dropped messages) should be considered and, if relevant, logged or surfaced to the user.
  - Testing
        While the test suite is updated and snapshots are refreshed, it would be good to add tests specifically targeting the new Map-based behavior (e.g., ensuring correct lookups, insertions, deletions in Maps).
  - Documentation
        The PR changes the shape of public types and APIs (e.g., in Instrument, component props). Consider updating any relevant documentation or README files.
        Adding JSDoc comments for new types and updated functions would help future maintainers.
  - Minor Issues
        In some test files, unused imports may remain (e.g., IfcBlock, IfcPV). Consider cleaning these up.
        Some string concatenations (e.g., when looking up keys in Maps) could benefit from helper functions to reduce the risk of typos.

## Notable Improvements

  - Dashboard and Run Info Rendering: The new approach in TopBar is more explicit about what is rendered and in what order, improving clarity.
  - Component Purity: The use of memo and useCallback helps avoid unnecessary re-renders.

## Summary

This is a comprehensive, well-executed refactor that improves data access performance and code clarity by switching from arrays to Maps for group and block data. The changes are thorough, with all relevant parts of the codebase updated and tested. While there is some added complexity in rendering and state management, the benefits appear to outweigh the drawbacks.

## Recommendation:
Approve after a quick check on the points above, especially around Map usage in React state/props and any subtle issues with equality/shallow comparison that may arise.

If you’d like a focused review on a specific file or concern, let me know!

rerpha reviewed (approved):
the robot told me to do it
